### PR TITLE
umapinfo: finale handling improvements, cosmetic changes

### PR DESCRIPTION
**Pull request by rfomin**
Test WAD: [intertext.zip](https://github.com/user-attachments/files/18185773/intertext.zip)
This test WAD works incorrectly in DSDA-Doom and GZDoom, but works fine in the KEX port.

### CODE REVIEW

**Comment by Pedro-Beirao** [comment]
Doesnt this possibly break some wads? Probably worth breaking them anyway, but needed to mention it

Right now, this broken behaviour is the "correct" one, since the big ports do this. So if we want to fix it, we should atleast tell the other ports, and also fix it in dsda 😉.
Otherwise, new wads could come out that would straightaway not work on woof, cuz they were tested in dsda/gz

**Comment by rfomin** [comment]
> Doesnt this possibly break some wads?

Currently, using `intertext = clear` with `endcast`, `endbunny`, etc. in "big ports" would break it, so I don't think anyone is using it. I plan to do big demo tests, as the DECLARATE implementation is coming.

> So if we want to fix it, we should atleast tell the other ports, and also fix it in dsda

Agreed, I think we should at least fix DSDA-Doom and GZDoom.

**Comment by rfomin** [comment]
> While we are in this discussion, perhaps it may be a good time to consider another revision on the `UMAPINFO` spec to be clearer and more explicit about its expected behavior, as well?

Yes, there are plans to update UMAPINFO in the future standard.

> I noticed on launch was that post-MAP32 slots had behaved differently than I expected from my time using Woof, which is not something currently defined in the spec.

Quote from the spec:
> The map names are limited to the format of the currently loaded IWAD, i.e. Doom 2 only supports MAPxx entries and Doom 1 only ExMy entries. The numbers x and y can exceed their original limits, though, so MAP50, E5M6 or even MAP100 or E1M10 are valid map names for their respective game.

So post-MAP32 slots should work. Please file a new issue if there are bugs in Woof.

**Comment by elf-alchemist** [comment]
> So post-MAP32 slots should work. Please file a new issue if there are bugs in Woof.
Nevermind me, then, I just made a test WAD with maps 33-40. Strange, I was under the impression that Woof "looped" the default Doom 2 definitions after MAP32, MAP38 (sixth map) would have a `InterText = "The End"`, MAP43 (11th map) would also have such an InterText, and so on. Perhaps something from an earlier version that has since changed and I didn't notice? Either way, thanks for clarifying.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -27,6 +27,7 @@ set(WOOF_SOURCES
     g_game.c               g_game.h
     g_input.c              g_input.h
     g_nextweapon.c         g_nextweapon.h
+    g_umapinfo.c           g_umapinfo.h
```
Comment (fabiangreffrath) [inline src/CMakeLists.txt:30]: Please don't forget about README.md

**Comment by fabiangreffrath** [comment]
> Test WAD: [intertext.zip](https://github.com/user-attachments/files/18185773/intertext.zip) This test WAD works incorrectly in DSDA-Doom and GZDoom, but works fine in the KEX port.

Hm, didn't PrBoom+UM once define what's wrong or right concerning UMAPINFO?

**Comment by rfomin** [comment]
> Hm, didn't PrBoom+UM once define what's wrong or right concerning UMAPINFO?

We've already changed the `endcast`/`endbunny` behavior so it's different in PrBoom+ compared to Woof/DSDA-Doom. Try test WAD, it works differently between all ports. Interestingly, Eternity tries to emulate PrBoom+ bugs.

The game/demo ends after `endcast`/`endbunny`, so these changes would probably be demo compatible. There might be rare WADs where `endcast` is present with `intertext = clear` by mistake, because in DSDA-Doom it just goes to the next level.

File: src/CMakeLists.txt (hunk 1 above)
Comment (rfomin) [reply src/CMakeLists.txt:30]: We don't mention `u_mapinfo.*` in the readme. Maybe we can change the copyright since this file is almost completely rewritten.

File: src/g_umapinfo.c [hunk 2]
```diff
@@ -285,7 +270,7 @@ static void ParseStandardProperty(scanner_t *s, mapentry_t *mape)
         {
             if (!strcasecmp(SC_GetString(s), "clear"))
             {
-                ReplaceString(&mape->label, "-");
+                mape->flags |= MapInfo_LabelClear;
```
Comment (fabiangreffrath) [inline src/g_umapinfo.c:273]: I have never understood why we would have `char[]` arrays in the umapinfo structs and check their first char for being `-` and similar. Why don't we just have `char *` pointers and it pointing to `NULL` will mean "cleared" and it pointing to some allocated string would mean this string.

Comment (rfomin) [reply src/g_umapinfo.c:273]: In the case of the `label` field, we need a separate "clear" state to remove the default label (e.g. "MAP01:"), so I added the `MapInfo_LabelClear` flag. We need a lot of flags, so it's also cumbersome.
`char[]` arrays indicate lump names wich should be 8 charaters max.

Comment (rfomin) [reply src/g_umapinfo.c:273]: > What I'm asking myself is, why did we set `mape->label[0] = '-'` instead of just letting `mape->label = NULL`? This would be a much clearer distinction between an empty string (`""`) and a cleared string (`NULL`).

Because we need 3 states in UMAPINFO:
1. `label = "Level 1"`, map title: "Level 1: Entryway"
2. No label field, `label = NULL` - use default, map title: "MAP01: Entryway".
3. `label = clear` - remove the default label, map title: "Entryway".

In this PR I have replaced `'-'` strings with flags.

Comment (fabiangreffrath) [reply src/g_umapinfo.c:273]: I see. We could replace the third state by a pointer to an empty string, but I'll leave this up to you of course.

Comment (rfomin) [reply src/g_umapinfo.c:273]: I think pointer to empty string is not much different then `"-"` string. I added flags because it's more readable, but that's a matter of taste.

File: src/g_umapinfo.h [hunk 3]
```diff
@@ -18,11 +18,30 @@
...
+#ifndef G_UMAPINFO_H
+#define G_UMAPINFO_H
 
 #include "doomtype.h"
 
+typedef enum
+{
+    MapInfo_LabelClear = (1u << 0),
+
+    MapInfo_EndGameArt = (1u << 2),
```
Comment (fabiangreffrath) [inline src/g_umapinfo.h:30]: Well, that's a really cumbersome way to write `4`. 😁

Comment (rfomin) [reply src/g_umapinfo.h:24]: I like `(1u << 0), (1u << 1), ... (1u << 32)` for flags more than hex or decimal values.

**Comment by fabiangreffrath** [comment]
Could you print a debug message if the situation is detected in which all major ports lead to different results? Will be interesting for us devs to know once WAD authors come to complain that their episodes do not end as expected.

Just something like this:

https://github.com/fabiangreffrath/crispy-doom/blob/9317be884bb03aa2df15f830083fcd6f6f484a1d/src/doom/deh_thing.c#L239-L243

**Comment by rfomin** [comment]
> Could you print a debug message if the situation is detected in which all major ports lead to different results? Will be interesting for us devs to know once WAD authors come to complain that their episodes do not end as expected.

Well, in the test wad case, DSDA-Doom/Eternity would go to the next level, PrBoom+ would restart the level, GZDoom would end the game without cast. So, something like this:
```c
if (flags & MapInfo_InterTextClear && flags & MapInfo_EndGameCast)
{
   I_Printf(VB_INFO, "'endcast' and 'intertext' fields don't work together in some ports");
}
```
I don't think there would be complaints about this.

**Comment by fabiangreffrath** [comment]
> So, something like this:

That would be great, yes.

> I don't think there would be complaints about this.

But at least we know what's happening (even after some time has passed).

**Comment by rfomin** [comment]
> > So, something like this:
> 
> That would be great, yes.

Done in 388b151

File: src/g_umapinfo.c [hunk 4]
```diff
@@ -219,19 +211,6 @@ static char *ParseMultiString(scanner_t *s)
 {
     char *build = NULL;
 
-    if (SC_CheckToken(s, TK_Identifier))
```
Comment (fabiangreffrath) [inline src/g_umapinfo.c:222]: Why was this part removed?

Comment (rfomin) [reply src/g_umapinfo.c:222]: I replaced all `"-"` strings with flags.

**Review by fabiangreffrath — APPROVED** [review]
Thank you!
